File: scripts/verify_setup.ts
/**
 * Verification script for TypeScript track setup.
 * Run this before Day 1 to ensure your environment is ready.
 */

import * as dotenv from 'dotenv';
import { StateGraph, START, END, StateSchema } from "@langchain/langgraph";
import { MemorySaver } from "@langchain/langgraph-checkpoint";
import { z } from "zod";

dotenv.config();

// interface is a structure that defines the properties of a type.
interface CheckResult {
  name: string;
  success: boolean;
  message: string;
}

// Promise is a type that represents the result of a asynchronous operation.
// the <> after Promise is a type parameter that specifies the type of the result of the asynchronous operation.
async function checkNodeVersion(): Promise<CheckResult> {
  const version = process.version;
  const major = parseInt(version.slice(1).split('.')[0]);
  
  if (major >= 18) {
    return {
      name: "Node Version",
      success: true,
      message: `✓ Node ${version}`
    };
  }
  return {
    name: "Node Version",
    success: false,
    message: `✗ Node ${version} (need 18+)`
  };
}

async function checkImports(): Promise<CheckResult> {
  try {
    // Already imported at top
    return {
      name: "Package Installation",
      success: true,
      message: "✓ LangGraph packages installed"
    };
  } catch (error: any) {
    return {
      name: "Package Installation",
      success: false,
      message: `✗ Missing packages: ${error.message}`
    };
  }
}

async function checkApiKeys(): Promise<CheckResult> {
  const openaiKey = process.env.OPENAI_API_KEY;
  const anthropicKey = process.env.ANTHROPIC_API_KEY;
  
  if (openaiKey || anthropicKey) {
    const providers = [];
    if (openaiKey) providers.push("OpenAI");
    if (anthropicKey) providers.push("Anthropic");
    return {
      name: "API Keys",
      success: true,
      message: `✓ API keys configured: ${providers.join(', ')}`
    };
  }
  return {
    name: "API Keys",
    success: false,
    message: "✗ No API keys found in .env"
  };
}

async function checkLangSmith(): Promise<CheckResult> {
  const langsmithKey = process.env.LANGSMITH_API_KEY;
  
  if (langsmithKey) {
    return {
      name: "LangSmith Config",
      success: true,
      message: "✓ LangSmith configured"
    };
  }
  return {
    name: "LangSmith Config",
    success: false,
    message: "⚠ LangSmith not configured (optional for Day 1-4)"
  };
}

async function testSimpleGraph(): Promise<CheckResult> {
  try {
    const State = new StateSchema({
      value: z.number(),
    });
    
    function node(state: typeof State.State) {
      return { value: state.value + 1 };
    }
    
    const workflow = new StateGraph(State)
      .addNode("test", node)
      .addEdge(START, "test")
      .addEdge("test", END);
    
    const graph = workflow.compile();
    const result = await graph.invoke({ value: 0 });
    
    if (result.value === 1) {
      return {
        name: "Simple Graph",
        success: true,
        message: "✓ Simple graph execution: SUCCESS"
      };
    }
    return {
      name: "Simple Graph",
      success: false,
      message: "✗ Graph returned unexpected result"
    };
  } catch (error: any) {
    return {
      name: "Simple Graph",
      success: false,
      message: `✗ Graph execution failed: ${error.message}`
    };
  }
}

async function testCheckpointer(): Promise<CheckResult> {
  try {
    const State = new StateSchema({
      count: z.number(),
    });
    
    function increment(state: typeof State.State) {
      return { count: state.count + 1 };
    }
    
    const checkpointer = new MemorySaver();
    const workflow = new StateGraph(State)
      .addNode("inc", increment)
      .addEdge(START, "inc")
      .addEdge("inc", END);
    
    const graph = workflow.compile({ checkpointer });
    
    const config = { configurable: { thread_id: "test" } };
    await graph.invoke({ count: 0 }, config);
    const result2 = await graph.invoke({ count: 0 }, config);
    
    if (result2.count === 2) {
      return {
        name: "Checkpointer",
        success: true,
        message: "✓ Checkpointer working: SUCCESS"
      };
    }
    return {
      name: "Checkpointer",
      success: false,
      message: "✗ Checkpointer not persisting state"
    };
  } catch (error: any) {
    return {
      name: "Checkpointer",
      success: false,
      message: `✗ Checkpointer test failed: ${error.message}`
    };
  }
}

async function testLlmConnection(): Promise<CheckResult> {
  const openaiKey = process.env.OPENAI_API_KEY;
  const anthropicKey = process.env.ANTHROPIC_API_KEY;
  
  if (!openaiKey && !anthropicKey) {
    return {
      name: "LLM Connection",
      success: false,
      message: "⚠ Skip LLM test (no API keys)"
    };
  }
  
  try {
    if (openaiKey) {
      const { ChatOpenAI } = await import("@langchain/openai");
      const model = new ChatOpenAI({ model: "gpt-4o-mini" });
      const response = await model.invoke([{ role: "user", content: "Hi" }]);
      if (response.content) {
        return {
          name: "LLM Connection",
          success: true,
          message: "✓ OpenAI API connection: SUCCESS"
        };
      }
    } else if (anthropicKey) {
      const { ChatAnthropic } = await import("@langchain/anthropic");
      const model = new ChatAnthropic({ model: "claude-3-5-haiku-20241022" });
      const response = await model.invoke([{ role: "user", content: "Hi" }]);
      if (response.content) {
        return {
          name: "LLM Connection",
          success: true,
          message: "✓ Anthropic API connection: SUCCESS"
        };
      }
    }
  } catch (error: any) {
    return {
      name: "LLM Connection",
      success: false,
      message: `⚠ LLM API test failed: ${error.message}`
    };
  }
  
  return {
    name: "LLM Connection",
    success: false,
    message: "⚠ LLM test inconclusive"
  };
}

async function main() {
  console.log("=".repeat(60));
  console.log("LangGraph Workshop - TypeScript Setup Verification");
  console.log("=".repeat(60));
  console.log();
  
  const checks = [
    checkNodeVersion,
    checkImports,
    checkApiKeys,
    checkLangSmith,
    testSimpleGraph,
    testCheckpointer,
    testLlmConnection,
  ];
  
  const results: CheckResult[] = [];
  
  for (const checkFn of checks) {
    const result = await checkFn();
    process.stdout.write(`Checking ${result.name}... `); // write to the console without a new line
    console.log(result.message);
    results.push(result);
  }
  
  console.log();
  console.log("=".repeat(60));
  
  const failed = results.filter(r => !r.success && !r.message.startsWith("⚠"));
  const warnings = results.filter(r => r.message.startsWith("⚠"));
  
  if (failed.length > 0) {
    console.log(`❌ ${failed.length} check(s) failed:`);
    failed.forEach(r => {
      console.log(`   - ${r.name}: ${r.message}`);
    });
    console.log();
    console.log("Please fix the issues above before starting the workshop.");
    console.log("See SETUP.md for detailed instructions.");
    process.exit(1);
  } else if (warnings.length > 0) {
    console.log(`⚠️  ${warnings.length} warning(s):`);
    warnings.forEach(r => {
      console.log(`   - ${r.name}: ${r.message}`);
    });
    console.log();
    console.log("You can proceed, but some features may not work.");
    console.log("All checks will be required by Day 5.");
  } else {
    console.log("✅ All checks passed! You're ready to start.");
  }
  
  console.log("=".repeat(60));
}

main().catch(console.error);

File: scripts/verify_setup.py
"""
Verification script for Python track setup.
Run this before Day 1 to ensure your environment is ready.
"""

import sys
import os
from typing import Dict, List, Tuple

def check_python_version() -> Tuple[bool, str]:
    """Check Python version is 3.11+."""
    version = sys.version_info
    if version >= (3, 11):
        return True, f"✓ Python {version.major}.{version.minor}.{version.micro}"
    return False, f"✗ Python {version.major}.{version.minor} (need 3.11+)"

def check_imports() -> Tuple[bool, str]:
    """Check required packages are installed."""
    try:
        import langgraph
        import langchain_core
        from langgraph.checkpoint.sqlite import SqliteSaver
        
        return True, f"✓ LangGraph {langgraph.__version__} installed"
    except ImportError as e:
        return False, f"✗ Missing package: {e.name}"

def check_api_keys() -> Tuple[bool, str]:
    """Check API keys are configured."""
    from dotenv import load_dotenv
    load_dotenv()
    
    openai_key = os.getenv("OPENAI_API_KEY")
    anthropic_key = os.getenv("ANTHROPIC_API_KEY")
    
    if openai_key or anthropic_key:
        providers = []
        if openai_key:
            providers.append("OpenAI")
        if anthropic_key:
            providers.append("Anthropic")
        return True, f"✓ API keys configured: {', '.join(providers)}"
    return False, "✗ No API keys found in .env"

def check_langsmith() -> Tuple[bool, str]:
    """Check LangSmith configuration."""
    from dotenv import load_dotenv
    load_dotenv()
    
    langsmith_key = os.getenv("LANGSMITH_API_KEY")
    if langsmith_key:
        return True, "✓ LangSmith configured"
    return False, "⚠ LangSmith not configured (optional for Day 1-4)"

def test_simple_graph() -> Tuple[bool, str]:
    """Test creating and running a simple graph."""
    try:
        from typing import TypedDict
        from langgraph.graph import StateGraph, START, END
        
        class State(TypedDict):
            value: int
        
        def node(state: State) -> dict:
            return {"value": state["value"] + 1}
        
        workflow = StateGraph(State)
        workflow.add_node("test", node)
        workflow.add_edge(START, "test")
        workflow.add_edge("test", END)
        
        graph = workflow.compile()
        result = graph.invoke({"value": 0})
        
        if result["value"] == 1:
            return True, "✓ Simple graph execution: SUCCESS"
        return False, "✗ Graph returned unexpected result"
        
    except Exception as e:
        return False, f"✗ Graph execution failed: {str(e)}"

def test_checkpointer() -> Tuple[bool, str]:
    """Test checkpointer functionality."""
    try:
        from typing import TypedDict
        from langgraph.graph import StateGraph, START, END
        from langgraph.checkpoint.memory import MemorySaver
        
        class State(TypedDict):
            count: int
        
        def increment(state: State) -> dict:
            return {"count": state["count"] + 1}
        
        workflow = StateGraph(State)
        workflow.add_node("inc", increment)
        workflow.add_edge(START, "inc")
        workflow.add_edge("inc", END)
        
        checkpointer = MemorySaver()
        graph = workflow.compile(checkpointer=checkpointer)
        
        config = {"configurable": {"thread_id": "test"}}
        result1 = graph.invoke({"count": 0}, config)
        result2 = graph.invoke({"count": 0}, config)
        
        if result2["count"] == 2:  # Should accumulate
            return True, "✓ Checkpointer working: SUCCESS"
        return False, "✗ Checkpointer not persisting state"
        
    except Exception as e:
        return False, f"✗ Checkpointer test failed: {str(e)}"

def test_llm_connection() -> Tuple[bool, str]:
    """Test LLM API connection (optional)."""
    from dotenv import load_dotenv
    load_dotenv()
    
    openai_key = os.getenv("OPENAI_API_KEY")
    anthropic_key = os.getenv("ANTHROPIC_API_KEY")
    
    if not (openai_key or anthropic_key):
        return False, "⚠ Skip LLM test (no API keys)"
    
    try:
        if openai_key:
            from langchain_openai import ChatOpenAI
            model = ChatOpenAI(model="gpt-4o-mini")
            response = model.invoke([{"role": "user", "content": "Hi"}])
            if response.content:
                return True, "✓ OpenAI API connection: SUCCESS"
        elif anthropic_key:
            from langchain_anthropic import ChatAnthropic
            model = ChatAnthropic(model="claude-3-5-haiku-20241022")
            response = model.invoke([{"role": "user", "content": "Hi"}])
            if response.content:
                return True, "✓ Anthropic API connection: SUCCESS"
    except Exception as e:
        return False, f"⚠ LLM API test failed: {str(e)}"

def main():
    """Run all verification checks."""
    print("=" * 60)
    print("LangGraph Workshop - Python Setup Verification")
    print("=" * 60)
    print()
    
    checks = [
        ("Python Version", check_python_version),
        ("Package Installation", check_imports),
        ("API Keys", check_api_keys),
        ("LangSmith Config", check_langsmith),
        ("Simple Graph", test_simple_graph),
        ("Checkpointer", test_checkpointer),
        ("LLM Connection", test_llm_connection),
    ]
    
    results = []
    for name, check_fn in checks:
        print(f"Checking {name}...", end=" ")
        success, message = check_fn()
        results.append((name, success, message))
        print(message)
    
    print()
    print("=" * 60)
    
    failed = [r for r in results if not r[1] and not r[2].startswith("⚠")]
    warnings = [r for r in results if r[2].startswith("⚠")]
    
    if failed:
        print(f"❌ {len(failed)} check(s) failed:")
        for name, _, msg in failed:
            print(f"   - {name}: {msg}")
        print()
        print("Please fix the issues above before starting the workshop.")
        print("See SETUP.md for detailed instructions.")
        sys.exit(1)
    elif warnings:
        print(f"⚠️  {len(warnings)} warning(s):")
        for name, _, msg in warnings:
            print(f"   - {name}: {msg}")
        print()
        print("You can proceed, but some features may not work.")
        print("All checks will be required by Day 5.")
    else:
        print("✅ All checks passed! You're ready to start.")
    
    print("=" * 60)

if __name__ == "__main__":
    main()
